### PR TITLE
[DAG] visitEXTRACT_VECTOR_ELT - constant fold legal fp imm values

### PR DESCRIPTION
**Pull request by RKSimon**
If we're extracting a constant floating point value, and the constant is a legal fp imm value, then replace the extraction with a fp constant.

### CODE REVIEW

File: llvm/test/CodeGen/X86/nontemporal-3.ll [hunk 1]
```diff
@@ -508,347 +327,86 @@ define void @test_zero_v8f64_align1(ptr %dst) nounwind {
...
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    xorl %eax, %eax
+; CHECK-NEXT:    movntiq %rax, 8(%rdi)
+; CHECK-NEXT:    movntiq %rax, (%rdi)
+; CHECK-NEXT:    movntiq %rax, 24(%rdi)
+; CHECK-NEXT:    movntiq %rax, 16(%rdi)
+; CHECK-NEXT:    movntiq %rax, 40(%rdi)
+; CHECK-NEXT:    movntiq %rax, 32(%rdi)
+; CHECK-NEXT:    movntiq %rax, 56(%rdi)
+; CHECK-NEXT:    movntiq %rax, 48(%rdi)
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/nontemporal-3.ll:340]: Is it less efficient than 4 x `movntsd %xmm0, xxxx`?

File: llvm/test/CodeGen/X86/2012-07-10-extload64.ll [hunk 2]
```diff
@@ -29,8 +29,8 @@ define void @store_64(ptr %ptr) {
 ; X86-LABEL: store_64:
 ; X86:       # %bb.0: # %BB
 ; X86-NEXT:    movl {{[0-9]+}}(%esp), %eax
-; X86-NEXT:    xorps %xmm0, %xmm0
-; X86-NEXT:    movlps %xmm0, (%eax)
+; X86-NEXT:    movl $0, 4(%eax)
+; X86-NEXT:    movl $0, (%eax)
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/2012-07-10-extload64.ll:33]: I wonder whether 2 `movl` is better than one `movlps`

File: llvm/test/CodeGen/X86/nontemporal-3.ll (hunk 1 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/nontemporal-3.ll:340]: movntsd are 64-bit nt stores so there would be 8 of them?

Comment (phoebewang) [reply llvm/test/CodeGen/X86/nontemporal-3.ll:340]: Oh, I see. I though they have 16B offset so should write 16B in a whole.

File: llvm/test/CodeGen/X86/2012-07-10-extload64.ll (hunk 2 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/2012-07-10-extload64.ll:33]: Looking at this now

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 3]
```diff
@@ -22243,6 +22243,19 @@ SDValue DAGCombiner::visitEXTRACT_VECTOR_ELT(SDNode *N) {
   unsigned NumElts = VecVT.getVectorNumElements();
   unsigned VecEltBitWidth = VecVT.getScalarSizeInBits();
 
+  // See if the extracted element is constant, in which case fold it if its
+  // a legal fp immediate.
+  if (IndexC && ScalarVT.isFloatingPoint()) {
+    APInt EltMask = APInt::getOneBitSet(NumElts, IndexC->getZExtValue());
+    KnownBits KnownElt = DAG.computeKnownBits(VecOp, EltMask);
+    if (KnownElt.isConstant()) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:22251]: Why is this fancy instead of just using dyn_cast<ConstantSDNode>?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:22251]: Its to handle bitcasts of vector data (e.g. <2 x i32>), which don't get folded away if we have multiple uses

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**